File: cpp/179_LargestNumber.cpp
class Solution {
public:
    string largestNumber(vector<int>& nums) {
        if (nums.empty()) 
            return "0";
        
        vector<string> num_strings;
        for (int n : nums)
            num_strings.push_back(to_string(n));
        
        sort(num_strings.begin(), num_strings.end(), [](const string& a, const string& b) {
            return a + b > b + a;
        });
        
        stringstream ss;
        for (string num_str : num_strings)
            ss << num_str;
        
        string result = ss.str();
        return result[0] == '0' ? "0" : result;
    }
};

File: java/179_LargestNumber.java
class Solution {
    private class LargerNumberComparator implements Comparator<String> {
        @Override
        public int compare(String a, String b) {
            String order1 = a + b;
            String order2 = b + a;
           return order2.compareTo(order1);
        }
    }

    public String largestNumber(int[] nums) {
        String[] numStrings = new String[nums.length];
        
        for (int i = 0; i < nums.length; ++i)
            numStrings[i] = String.valueOf(nums[i]);
        
        Arrays.sort(numStrings, new LargerNumberComparator());
        if (numStrings[0].equals("0"))
            return "0";
        
        StringBuilder result = new StringBuilder();
        for (String s : numStrings)
            result.append(s);
        
        return result.toString();
    }
}
